### DISCUSSION TITLE
How to use roll_time_series function properly?

### DISCUSSION THREAD

**Discussion by Mahyar24** [blue-yonder/tsfresh#1037] [Q&A] [answered]
Hi,

I have a time-series dataset with minute-by-minute data. I would like to get the last 4 hours of data to predict a classification label for the next **1-hour**. To achieve this, I created IDs using `pd.Grouper(freq="4H")` to group **every 4-hours** together. However, when I don't use the `roll_time_series` function, the resulting tabularized dataset will look like this:

```
2023-01-01 00:00:00 -> 2023-01-01 04:00:00
2023-01-01 04:00:00 -> 2023-01-01 08:00:00
```

Unfortunately, in this format, I am losing lots of data. What I actually want is:

```
2023-01-01 00:00:00 -> 2023-01-01 04:00:00
2023-01-01 01:00:00 -> 2023-01-01 05:00:00
```

I'm not sure how to create it accordingly with the `roll_time_series` function. Perhaps it should involve using parameters such as `max_timeshift` and `min_timeshift`, but I have no clue what values these parameters should take.

P.S. Here's an example of my data structure:

| date                | feature_1 | feature_2 | id             |
| ------------------- | --------- | --------- | -------------- |
| **2023-01-21 00:00:00**          | -0.260733   | -9999     | **20230121000000** |
| **2023-01-21 00:01:00**          | -0.263591   | -9999     | **20230121000000** |

**Comment by nils-braun** [comment, marked as answer]
Have you looked into the documentation for the function? https://tsfresh.readthedocs.io/en/latest/api/tsfresh.utilities.html#tsfresh.utilities.dataframe_functions.roll_time_series
There is also a dedicated document text for rolling: https://tsfresh.readthedocs.io/en/latest/text/forecasting.html#forecasting-label

There you will find what these arguments mean. For your use-case, you can set the `max_timeshift` to 4*60 (maximum length of each resulting time series) and the `rolling_direction` to 60 (the size of each step). Have you actually tried the function already and are facing any errors with it?

**Reply by Mahyar24** [reply]
Yep, I've read it and appreciate it. IMHO, however, it isn't clear enough (at least for me!). I was tinkering with `max_timeshift` and `min_timeshift`, but they didn't produce the desired outcome. Additionally, I ignored `rolling_direction` because, based on the variable name, I thought it was irrelevant to my purpose. Sorry about that and thanks.

**Reply by nils-braun** [reply]
No worries - it's good you asked here! Do you have a suggestion on how to improve it? Feel free to do a PR or tell us about it :) Happy to incorporate your feedback!

**Reply by Mahyar24** [reply]
It's hard for me to wrap my mind around the source code, and eventually, I had to write a helper function for myself. However, I highly recommend renaming `rolling_direction` to something like `steps` and using annotations corresponding to variable names on the images in the docs. It makes it so much easier to understand.
